### ISSUE TITLE
ES6 module help please

### ISSUE THREAD

**Issue by flatmax** [emscripten-core/emscripten#13754]
I am trying to load an ES6 module into the browser, but having some difficulty. I can't find anything on the net about it ...

I am binding an ES6 module like so :
	@emcc --bind --llvm-opts 1 --memory-init-file 0 --pre-js ../webcomponents/pre.js -s EXPORT_ES6=1 -s MODULARIZE=1 -s EXPORT_NAME="'libgtkIOStream'" -s SINGLE_FILE=1 -s "BINARYEN_METHOD='native-wasm'" -s WASM=1 -s ALLOW_MEMORY_GROWTH=1 $(AM_CXXFLAGS) .libs/libgtkIOStream.so $(SOX_LIBS) -o .libs/libgtkIOStream.js

pre.js looks like so :
```
Module['onRuntimeInitialized'] = function(){
  window.dispatchEvent(new CustomEvent('libgtkIOStream::initialised'));
  console.log('WASM initalised')
};
```
I instantiate the module in the browser like so :
```
export var libgtkIOStream;
import("./libgtkIOStream").then((mod)=>{
  libgtkIOStream=mod;
  console.log(mod);
})
```
However the function onRuntimeInitialized is never called.

The contents of the module don't look as I would expect. console.log(mod) give the following :
```
Module {Symbol(Symbol.toStringTag): "Module"}
default: (...)
Symbol(Symbol.toStringTag): "Module"
get default: ƒ ()
set default: ƒ ()
```

Any pointers on what I am doing wrong ?

**Comment by sbc100** [comment]
First of all you don't need ` --llvm-opts 1 -s "BINARYEN_METHOD='native-wasm'" -s WASM=1` anymore.  If you found those in tutorial or Makefile can you please update to link to it so that we can find a way to update.   You probably also don't need/want `-s SINGLE_FILE=1`.      Removing unneeded options like that moves you towards a more well lit path where things should be easier to debug.  Thats just general advice though... in terms of your actual issue I'm not sure :(   @RReverser ?

**Comment by RReverser** [comment]
I mean... you're not doing anything wrong. The Module object looks as expected for ES6 output. The default export you're seeing is the factory function that you pass options into, and get result back.

Check e.g. https://developers.google.com/web/updates/2019/01/emscripten-npm for usage examples.

**Comment by RReverser** [comment]
Or, rather, let me rephrase, what's wrong is this loading code:

```js
export var libgtkIOStream;
import("./libgtkIOStream").then((mod)=>{
  libgtkIOStream=mod;
  console.log(mod);
})
```

It needs to instantiate an actual module using `mod.default(...options...)`, `mod` itself is just ES6 module object, not Emscripten's `Module` (naming is hard...).

**Comment by RReverser** [comment]
Btw, generally it's better to ask questions on StackOverflow - there is a larger community that can help with questions until you're sure you're hitting a bug.

I'll close this as this doesn't seem like an issue.

**Comment by flatmax** [comment]
Ok, worked it out ...
https://github.com/flatmax/gtkiostream/blob/master/webcomponents/gtkiostream/libgtkiostream.js#L2-L7

```
import modProm from './libgtkIOStream.js';
export var libgtkIOStream;
modProm().then((mod)=>{
  libgtkIOStream = mod;
  window.dispatchEvent(new CustomEvent('libgtkIOStream::initialised'));
})
```
